From c9a877ea48935bb40597cc2e7d88b0035a27e6f2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Joan=20Bruguera=20Mic=C3=B3?= <joanbrugueram@gmail.com>
Date: Sat, 12 Apr 2025 22:46:59 +0000
Subject: [PATCH] Tentative fix for NVIDIA 470.256.02 driver for Linux 6.15-rc1

The trickier part of this patch is that vm_flags_set and vm_flags_clear
depend on a GPL-only symbol (__vma_start_write) since Linux >=6.15.

In some situations (mainly nvidia_mmap_helper), we can replace it by
vm_flags_reset as the VMA gets locked by other previous calls.
In some other code paths that I have not been able to trigger, I don't
believe the VMA is locked, so it returns -ENOTSUPP instead for now.

Thanks to Willy Frissen and Satadru Pramanik, who shared patches for
various of the necessary changes.
---
 Kbuild                                  | 25 +++++++++----------------
 common/inc/nv-mm.h                      | 18 ++++++++++++++++++
 common/inc/nv-timer.h                   |  1 +
 nvidia-drm/nvidia-drm-connector.c       |  8 ++++++++
 nvidia-drm/nvidia-drm-gem-user-memory.c |  9 +++++++++
 nvidia-drm/nvidia-drm-gem.c             |  5 +++++
 nvidia-drm/nvidia-drm-linux.c           |  1 +
 nvidia-modeset/nvidia-modeset-linux.c   |  5 +++--
 nvidia-uvm/uvm.c                        | 10 ++++++++++
 nvidia/nv-frontend.c                    |  1 +
 nvidia/nv-mmap.c                        |  6 ++++++
 nvidia/nv.c                             |  4 ++--
 12 files changed, 73 insertions(+), 20 deletions(-)

diff --git a/common/inc/nv-mm.h b/common/inc/nv-mm.h
index da5065d..094cd8a 100644
--- a/common/inc/nv-mm.h
+++ b/common/inc/nv-mm.h
@@ -31,6 +31,7 @@ typedef int vm_fault_t;
 
 #include <linux/mm.h>
 #include <linux/sched.h>
+#include <linux/version.h>
 /* get_user_pages
  *
  * The 8-argument version of get_user_pages was deprecated by commit
@@ -248,12 +249,29 @@ static inline struct rw_semaphore *nv_mmap_get_lock(struct mm_struct *mm)
 #if defined(NV_VM_AREA_STRUCT_HAS_CONST_VM_FLAGS)
 static inline void nv_vm_flags_set(struct vm_area_struct *vma, vm_flags_t flags)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 15, 0)
+    // Rel. commit "mm: uninline the main body of vma_start_write()" (Suren Baghdasaryan, 13 Feb 2025)
+    // Since Linux 6.15, vm_flags_set and vm_flags_clear call a GPL-only symbol
+    // for locking (__vma_start_write), which can't be called from non-GPL code.
+    // However, it appears all uses on the driver are on VMAs being initially
+    // mapped and which are already locked by previous calls over that VMA,
+    // so we can use vm_flags_reset, which doesn't lock the VMA, but rather just
+    // asserts it is already write-locked.
+    vm_flags_reset(vma, vma->vm_flags | flags);
+#else
     vm_flags_set(vma, flags);
+#endif
 }
 
 static inline void nv_vm_flags_clear(struct vm_area_struct *vma, vm_flags_t flags)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 15, 0)
+    // Rel. commit "mm: uninline the main body of vma_start_write()" (Suren Baghdasaryan, 13 Feb 2025)
+    // See above
+    vm_flags_reset(vma, vma->vm_flags & ~flags);
+#else
     vm_flags_clear(vma, flags);
+#endif
 }
 #else
 static inline void nv_vm_flags_set(struct vm_area_struct *vma, unsigned long flags)
diff --git a/nvidia-drm/nvidia-drm-gem-user-memory.c b/nvidia-drm/nvidia-drm-gem-user-memory.c
index 6de9dd9..e56efb0 100644
--- a/nvidia-drm/nvidia-drm-gem-user-memory.c
+++ b/nvidia-drm/nvidia-drm-gem-user-memory.c
@@ -36,6 +36,8 @@
 #include "linux/mm.h"
 #include "nv-mm.h"
 
+#include <linux/version.h>
+
 static inline
 void __nv_drm_gem_user_memory_free(struct nv_drm_gem_object *nv_gem)
 {
@@ -92,6 +94,13 @@ static int __nv_drm_gem_user_memory_mmap(struct nv_drm_gem_object *nv_gem,
         return -EINVAL;
     }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 15, 0)
+    // Rel. commit "mm: uninline the main body of vma_start_write()" (Suren Baghdasaryan, 13 Feb 2025)
+    printk(KERN_WARNING "[NVIDIA] Unable to call vm_flags_(set|clear) on "
+           "%s on Linux >=6.15, for more information see %s.\n", __func__,
+           "https://gist.github.com/joanbm/def2bf57ed7a799c1d84a67606459314");
+    return -ENOTSUPP;
+#endif
     nv_vm_flags_clear(vma, VM_PFNMAP);
     nv_vm_flags_clear(vma, VM_IO);
     nv_vm_flags_set(vma, VM_MIXEDMAP);
diff --git a/nvidia-drm/nvidia-drm-gem.c b/nvidia-drm/nvidia-drm-gem.c
index f7b4b3f..22ec4bf 100644
--- a/nvidia-drm/nvidia-drm-gem.c
+++ b/nvidia-drm/nvidia-drm-gem.c
@@ -49,6 +49,8 @@
 
 #include "nv-mm.h"
 
+#include <linux/version.h>
+
 void nv_drm_gem_free(struct drm_gem_object *gem)
 {
     struct nv_drm_gem_object *nv_gem = to_nv_gem_object(gem);
@@ -297,6 +299,9 @@ int nv_drm_mmap(struct file *file, struct vm_area_struct *vma)
             ret = -EINVAL;
             goto done;
         }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 15, 0)
+        #error "Not expecting to compile the vm_flags_clear call in nv_drm_mmap"
+#endif
         nv_vm_flags_clear(vma, VM_MAYWRITE);
     }
 #endif
diff --git a/nvidia-uvm/uvm.c b/nvidia-uvm/uvm.c
index 73ceb70..321a411 100644
--- a/nvidia-uvm/uvm.c
+++ b/nvidia-uvm/uvm.c
@@ -36,6 +36,8 @@
 #include "uvm_hmm.h"
 #include "uvm_mem.h"
 
+#include <linux/version.h>
+
 #define NVIDIA_UVM_DEVICE_NAME          "nvidia-uvm"
 
 static dev_t g_uvm_base_dev;
@@ -812,6 +814,13 @@ static int uvm_mmap(struct file *filp, struct vm_area_struct *vma)
     // Using VM_DONTCOPY would be nice, but madvise(MADV_DOFORK) can reset that
     // so we have to handle vm_open on fork anyway. We could disable MADV_DOFORK
     // with VM_IO, but that causes other mapping issues.
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 15, 0)
+    // Rel. commit "mm: uninline the main body of vma_start_write()" (Suren Baghdasaryan, 13 Feb 2025)
+    printk(KERN_WARNING "[NVIDIA] Unable to call vm_flags_(set|clear) on "
+           "%s on Linux >=6.15, for more information see %s.\n", __func__,
+           "https://gist.github.com/joanbm/def2bf57ed7a799c1d84a67606459314");
+    return -ENOTSUPP;
+#endif
     nv_vm_flags_set(vma, VM_MIXEDMAP | VM_DONTEXPAND);
 
     vma->vm_ops = &uvm_vm_ops_managed;
diff --git a/nvidia/nv-mmap.c b/nvidia/nv-mmap.c
index 9e1eb5d..81bd7a6 100644
--- a/nvidia/nv-mmap.c
+++ b/nvidia/nv-mmap.c
@@ -15,6 +15,8 @@
 #include "nv-ibmnpu.h"
 #include "nv_speculation_barrier.h"
 
+#include <linux/version.h>
+
 /*
  * The 'struct vm_operations' open() callback is called by the Linux
  * kernel when the parent VMA is split or copied, close() when the
@@ -448,7 +450,11 @@ static int nvidia_mmap_numa(
     }
 
     // Needed for the linux kernel for mapping compound pages
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 15, 0)
+    // Rel. commit "mm: uninline the main body of vma_start_write()" (Suren Baghdasaryan, 13 Feb 2025)
+    // This call isn't necessary because vm_insert_page already sets VM_MIXEDMAP
     nv_vm_flags_set(vma, VM_MIXEDMAP);
+#endif
 
     for (i = 0, addr = mmap_context->page_array[0]; i < pages;
          addr = mmap_context->page_array[++i], start += PAGE_SIZE)
-- 
2.49.0

